File: desktop/test/unit/localEcho.test.ts
// Port of the Android LocalEchoTest: the judgment calls of optimistic echo.
// A wrong guess here is a ghost character floating in a live pane — worse than
// the latency it hides — so most of these tests are about NOT rendering.

import { describe, expect, it } from 'vitest'
import {
  backspace, emptyEcho, frame, MAX_PENDING, otherKey, typed, visible, type Echo,
} from '../../src/shared/core/localEcho'

const muted: Echo = { text: '', muted: true }

describe('LocalEcho', () => {
  it('typed characters accumulate and are visible', () => {
    let e = emptyEcho
    e = typed(e, 'l')
    e = typed(e, 's')
    expect(e.text).toBe('ls')
    expect(visible(e)).toBe(true)
  })

  it('backspace eats the last pending character', () => {
    let e = typed(emptyEcho, 'ab')
    e = backspace(e)
    expect(e.text).toBe('a')
  })

  it('backspace past the buffer mutes - the effect on screen is unknowable', () => {
    const e = backspace(emptyEcho)
    expect(e.muted).toBe(true)
    expect(visible(e)).toBe(false)
  })

  it('unpredictable keys mute until a frame settles things', () => {
    const e = otherKey(typed(emptyEcho, 'half'))
    expect(e.muted).toBe(true)
    expect(visible(e)).toBe(false)
  })

  it('typing while muted stays muted - no guessing on unknown ground', () => {
    const e = typed(muted, 'x')
    expect(e.muted).toBe(true)
    expect(e.text).toBe('')
  })

  it("a frame consumes exactly the cursor's advance", () => {
    const e = typed(emptyEcho, 'abc')
    // The pane confirmed two characters: cursor moved 2 cells on the same row.
    const after = frame(e, [10, 5], [12, 5])
    expect(after.text).toBe('c')
    expect(visible(after)).toBe(true)
  })

  it('a frame that consumes everything leaves nothing pending', () => {
    const e = typed(emptyEcho, 'ab')
    expect(frame(e, [4, 2], [6, 2]).text).toBe('')
  })

  it('a row change clears - wraps are exactly where ghosts come from', () => {
    const e = typed(emptyEcho, 'abcdef')
    expect(frame(e, [78, 5], [4, 6]).text).toBe('')
  })

  it('a backwards cursor clears - the pane redrew, all bets off', () => {
    const e = typed(emptyEcho, 'abc')
    expect(frame(e, [10, 5], [2, 5]).text).toBe('')
  })

  it('an advance longer than the buffer clears rather than inventing text', () => {
    const e = typed(emptyEcho, 'ab')
    expect(frame(e, [10, 5], [20, 5]).text).toBe('')
  })

  it('a frame lifts a mute - it is the resolution being waited for', () => {
    const e = frame(muted, [3, 1], [4, 1])
    expect(e.muted).toBe(false)
    expect(e.text).toBe('')
  })

  it('the first frame ever clears rather than guessing at history', () => {
    const e = typed(emptyEcho, 'abc')
    expect(frame(e, null, [5, 5]).text).toBe('')
  })

  it('a runaway buffer mutes instead of building a longer ghost', () => {
    let e = emptyEcho
    for (let i = 0; i < MAX_PENDING + 1; i++) e = typed(e, 'x')
    expect(e.muted).toBe(true)
  })
})

File: mobile/core/src/commonTest/kotlin/com/silencelen/huginn/LocalEchoTest.kt
package com.silencelen.huginn

import com.silencelen.huginn.ui.LocalEcho
import com.silencelen.huginn.ui.LocalEcho.Echo
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlin.test.Test

/**
 * The judgment calls of optimistic echo. A wrong guess here is a ghost
 * character floating in a live pane — worse than the latency it hides — so most
 * of these tests are about NOT rendering.
 */
class LocalEchoTest {

    @Test
    fun `typed characters accumulate and are visible`() {
        var e = Echo()
        e = LocalEcho.typed(e, "l")
        e = LocalEcho.typed(e, "s")
        assertEquals("ls", e.text)
        assertTrue(e.visible)
    }

    @Test
    fun `backspace eats the last pending character`() {
        var e = LocalEcho.typed(Echo(), "ab")
        e = LocalEcho.backspace(e)
        assertEquals("a", e.text)
    }

    @Test
    fun `backspace past the buffer mutes - the effect on screen is unknowable`() {
        val e = LocalEcho.backspace(Echo())
        assertTrue(e.muted)
        assertFalse(e.visible)
    }

    @Test
    fun `unpredictable keys mute until a frame settles things`() {
        val e = LocalEcho.otherKey(LocalEcho.typed(Echo(), "half"))
        assertTrue(e.muted)
        assertFalse(e.visible)
    }

    @Test
    fun `typing while muted stays muted - no guessing on unknown ground`() {
        val e = LocalEcho.typed(Echo(muted = true), "x")
        assertTrue(e.muted)
        assertEquals("", e.text)
    }

    @Test
    fun `a frame consumes exactly the cursor's advance`() {
        val e = LocalEcho.typed(Echo(), "abc")
        // The pane confirmed two characters: cursor moved 2 cells on the same row.
        val after = LocalEcho.frame(e, prev = 10 to 5, cur = 12 to 5)
        assertEquals("c", after.text)
        assertTrue(after.visible)
    }

    @Test
    fun `a frame that consumes everything leaves nothing pending`() {
        val e = LocalEcho.typed(Echo(), "ab")
        assertEquals("", LocalEcho.frame(e, 4 to 2, 6 to 2).text)
    }

    @Test
    fun `a row change clears - wraps are exactly where ghosts come from`() {
        val e = LocalEcho.typed(Echo(), "abcdef")
        assertEquals("", LocalEcho.frame(e, 78 to 5, 4 to 6).text)
    }

    @Test
    fun `a backwards cursor clears - the pane redrew, all bets off`() {
        val e = LocalEcho.typed(Echo(), "abc")
        assertEquals("", LocalEcho.frame(e, 10 to 5, 2 to 5).text)
    }

    @Test
    fun `an advance longer than the buffer clears rather than inventing text`() {
        val e = LocalEcho.typed(Echo(), "ab")
        assertEquals("", LocalEcho.frame(e, 10 to 5, 20 to 5).text)
    }

    @Test
    fun `a frame lifts a mute - it is the resolution being waited for`() {
        val e = LocalEcho.frame(Echo(muted = true), 3 to 1, 4 to 1)
        assertFalse(e.muted)
        assertEquals("", e.text)
    }

    @Test
    fun `the first frame ever clears rather than guessing at history`() {
        val e = LocalEcho.typed(Echo(), "abc")
        assertEquals("", LocalEcho.frame(e, prev = null, cur = 5 to 5).text)
    }

    @Test
    fun `a runaway buffer mutes instead of building a longer ghost`() {
        var e = Echo()
        repeat(LocalEcho.MAX_PENDING + 1) { e = LocalEcho.typed(e, "x") }
        assertTrue(e.muted)
    }
}
